### PR TITLE
feat: Add Compaction Regions

### PR DESCRIPTION
**Pull request by Joe-Blount**
helps: https://github.com/influxdata/idpe/issues/17974

This PR adds preservation of a previous `RoundInfo` for consideration in the next round of compaction.  At present, the only history acted on is a new concept called `ComapactRange`, where the `RoundInfo` computation identifies and remembers a previous vertical splitting.  `CompactRange`s continue to exist until all the `L0`s are compacted down to `L1`.  This means once we do a vertical split, we retain the knowledge of how it split the files, ensuring we can treat the resulting small units of work as separate small backlogs of `L0`s (which are efficiently handled).

**Commit 1** introduces `CompactRange` is the first commit in this PR.

**Commit 2**  shows the insta test updates for adding `CompactRange`.  Several tests show some nice improvements in write amplification, such as 717mb written decreasing to 240mb, a 66% improvement.  A few show significant worsening of write amplification and/or not completing compaction. None of these degredations remain in the final commit.  So ignore the second commit if you'd like, or look at it if your'e curious.

**Commit 3** Preserving vertical split results in `CompactRanges` means the scenarios `ManySmallFiles` is asked to handle are well defined and limited in size.  This allows a change (commit 3) which makes `ManySmallFiles` skip some larger `L0`s instead of potentially rewriting them many times.  Because `ManySmallFiles` knows the `L0` total isn't too big, it can compute an acceptable `L0` size that ensures it reduces to the desired file count.  Skipping these files allows `ManySmallFiles` to avoid repeatedly rewriting moderately sized `L0`s - which in some cases significantly helps write amplification.

**Commit 4** is insta updates for the `ManySmallFiles` change in commit 3.  The case that prompted me to make this change shows the most dramatic improvement, reducing from 24.48gb written to 7.22gb, a 70% reduction in bytes written, with identical final output from the compaction :partying_face: .

**Commit 5** reduces the threshold for upgrading files to 50% of the maximum file size, rather than 100%.  The max file size is a soft target that we rarely hit precisely.  So any file we've already grown to be up near that size is fine to just upgrade.  Note that on a gappy leading edge workload we'll be left with much smaller files because we don't rewrite them once they're 10% of the max.  So upgrading  at  50% threshold is plenty strict.

**Commit 6** is the inst updates for commit 5, showing more improvements in write amplification.

A very welcome side affect of adding the `CompactRegions` is that the long simulation tests are much easier to reason about.  Some have several hundred split|compact plans.  Scrolling through hundreds of operations and trying to identify a logical strategy is challenging.  With the `CompactRegions`, its fairly easy to search for the timestamp of a region and see the plans pertaining to it, which is a short list.  Preserving the region and considering it as its own small backlog to dealt with makes judging the 1-5 plans for an easy thing to do.

Commits 1,3,5 should be reviewed.
Commits 2,4,6 can be reviewed, or just look at the insta test updates in the final dif.

Every affected test case is either neutral or improved.  None are made worse by this PR.

Here's the overall summary list of changes to write amplification:
test name | written (orig) | written (final) | percent delta (improvement) | notes
--- | --- | --- | --- | --- 
random_backfill_empty_partition | 2.34gb | 1.37gb | 41.4 | its more compacted than before, a L1 previously left behind now isn't |
random_backfill_over_l2s | 4.04gb | 3.32gb | 17.8 | its less compacted than before, a few L1s are now left to accumulate more data before compaction |
actual_case_from_catalog_1 | 16.17gb | 14.58gb | 9.8 |
all_non_overlapping_l0 | 200mb | 120mb |40 |
l1_too_much_with_non_overlapping_l0 | 760mb | 310mb | x | (disregard, test changed to maintain scenario)
large_l1_with_non_overlapping_l0 |200mb | 115mb | x | (disregard, test changed to maintain scenario)
all_overlapping_l0_max_input_bytes_per_partition | 717mb | 240mb | 66.5 |
all_overlapping_l0_max_input_bytes_per_partition_small_max_desired_file_size | 717mb | 240mb | 66.5 |
two_large_files_total_over_max_compact_size_start_l0 | 800mb | 620mb | 22.5 |
many_good_size_l0_files | 1.22gb | 580mb | 53.6 |
l0s_needing_vertical_split | 4.32gb | 2.63gb | 39.1 |
 many_medium_files_time_range_1 | 5.42gb | 1.46gb | 73 |
stuck_l0 | 39.99gb | 26.51gb | 33.7 | its more compacted than before, a L1 previously left behind now isn't |
stuck_l0_large_l0s | 6.39gb | 5.87gb | 8.1  |
split_then_undo_it | 1.33gb | 975mb | 28.4 |
split_precent_loop | 2.78gb | 1.98gb | 28.8 |
very_big_overlapped_backlog | 40.58gb | 40.21gb | 0.9 |

In the last few weeks I've revived old highly backlogged partitions and watched them catch up.  As the compactor gets further behind, it is less efficient at catching up.  Anecdotally, it appears that as the backlog increases, the decrease in efficiency is exponential. Those scenarios are big enough it is challenging to capture them in manageable test cases.  I'm optimistic that this PR will provide significant improvement in the worst of those scenarios, even though we can't quantify the benefit numerically.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/components/round_info_source/mod.rs [hunk 1]
```diff
@@ -284,9 +293,64 @@ impl LevelBasedRoundInfo {
...
+            for chain in &chains {
+                let mut min = chain.iter().map(|f| f.min_time).min().unwrap();
+
+                let max = chain.iter().map(|f| f.max_time).max().unwrap();
+
+                if min <= prior_overlapping_max && prior_overlapping_max != Timestamp::new(0) {
+                    // Target level files overlap more than one start level file, and there is a target level file overlapping
+                    // the prior chain of L0s and this one.  We'll split the target level file at the before proceeding with
+                    // compactions.
+                    split_times.push(prior_chain_max)
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:312]: This is getting a bit hard to follow.

After the last PR, the split_times generated in this methods are hints based upon the density distribution. And later downstream code decides the actual split times based on the existing files (e.g. L1 file edges). ("hints" versus actual "times" decided.)

This change here further muddles things. We are now have the split_hints here be both density edges and range edges. Can we do something like the following?
```
let (mut split_hints, ranges) = self
    .chain_range_times(files.clone().to_vec(), self.max_total_file_size_per_plan);
split_hints.push(self
    .vertical_split_times(files.clone().to_vec(), self.max_total_file_size_per_plan));

if !split_hints.is_empty() {
    RoundInfo::VerticalSplit { split_times: split_hints }
} else if !ranges.is_empty() {
    RoundInfo::CompactRanges {
        ranges,
```

File: compactor/src/components/round_info_source/mod.rs [hunk 2]
```diff
@@ -284,9 +293,64 @@ impl LevelBasedRoundInfo {
...
+                    max: max.get(),
+                    cap,
+                });
+            }
+        }
+
         split_times.sort();
         split_times.dedup();
-        split_times
+        (split_times, ranges)
```
[the remark this answers is not in the document]
Comment (wiedld) [reply compactor/src/components/round_info_source/mod.rs:353]: See other comment -- can we break it up to something like `chain_range_times()` and `vertical_split_times()`?

File: compactor/src/components/round_info_source/mod.rs [hunk 3]
```diff
@@ -309,14 +394,21 @@ impl RoundInfoSource for LevelBasedRoundInfo {
...
+            let (split_times, ranges) = self
                 .vertical_split_times(files.clone().to_vec(), self.max_total_file_size_per_plan);
+
             if !split_times.is_empty() {
                 RoundInfo::VerticalSplit { split_times }
-            } else if self.many_ungroupable_files(&files, start_level, self.max_num_files_per_plan)
-            {
-                RoundInfo::SimulatedLeadingEdge {
+            } else if !ranges.is_empty() {
+                RoundInfo::CompactRanges {
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:410]: Since we now have `RoundInfo::SimulatedLeadingEdge` as dead code, can we add the removal of this to our tasks list?

File: compactor/src/components/file_classifier/split_based.rs [hunk 4]
```diff
@@ -207,6 +207,74 @@ where
...
+                    .filter(|f| f.compaction_level == CompactionLevel::Initial)
+                    .count();
+
+                if l0_count > *max_num_files_to_group {
+                    // Too many L0s, do manySmallFiles within this range.
+                    let (files_to_compact, mut files_to_keep) = files_to_compact
+                        .into_iter()
+                        .partition(|f| f.compaction_level == CompactionLevel::Initial);
+
+                    let l0_classification = file_classification_for_many_files(
```
Comment (wiedld) [inline compactor/src/components/file_classifier/split_based.rs:227]: In the round calculation, we have a clear delineation of:
`RoundInfo::CompactRanges` versus `RoundInfo::ManySmallFiles`

Here we have downstream code which generates a compact_reason of ManySmallFiles.
=> this code here will output file_classifier with compact_reason `CompactReason::ManySmallFile`.

This feels confusing?

The rest of the compact reasons are:
```
pub enum CompactReason {
    ManySmallFiles,
    TotalSizeLessThanMaxCompactSize,
    FoundSubsetLessThanMaxCompactSize,
}
```

Maybe we need to rename `RoundInfo::ManySmallFiles`, and get better at defining what it is?

File: compactor/src/components/round_info_source/mod.rs [hunk 5]
```diff
@@ -309,14 +394,21 @@ impl RoundInfoSource for LevelBasedRoundInfo {
             self.max_total_file_size_per_plan,
         );
 
-        let round_info = if start_level == CompactionLevel::Initial {
-            let split_times = self
+        let round_info = if !ranges.is_empty() {
+            RoundInfo::CompactRanges {
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:398]: This would be less confusing, if we renamed this to `ranges_still_not_compacted`.
(Since these are the remaining ranges from last round.)

Also, why do we have uncompacted ranges from last round? Is this because the last round's `RoundInfo::CompactRanges` then had files removed during the downstream file classifier code? Isn't this part of the split brain problem?

File: compactor/src/components/split_or_compact/start_level_files_to_split.rs [hunk 6]
```diff
@@ -358,6 +359,9 @@ pub fn merge_small_l0_chains(
         }
     }
 
+    // Put it back in the standard order.
+    merged_chains.sort_by_key(|a| a[0].min_time);
```
Comment (wiedld) [inline compactor/src/components/split_or_compact/start_level_files_to_split.rs:363]: Ah, this looks like an existing bug. 
Maybe in it's own PR, so we can parse out what impacts on compaction are from this change?

File: compactor/tests/integration.rs [hunk 7]
```diff
@@ -58,7 +58,7 @@ async fn test_num_files_over_limit() {
     //
     // read files and verify 2 files
     let files = setup.list_by_table_not_to_delete().await;
-    assert_eq!(files.len(), 2);
+    assert_eq!(files.len(), 3);
```
Comment (wiedld) [inline compactor/tests/integration.rs:61]: The comment two lines above still reads `verify 2 files`.

File: compactor/src/components/file_classifier/split_based.rs [hunk 8]
```diff
@@ -280,6 +280,9 @@ where
 }
 
 // ManySmallFiles assumes the L0 files are tiny and aims to do L0-> L0 compaction to reduce the number of tiny files.
+// With vertical splitting, this only operates on a CompactionRange that's up to the max_compact_size.  So while there's
+// many files, we know there's not many bytes (in total).  Because of this, We can skip anything that's a sizeable portion
+// of the max_compact_size, knowing that we can still get the L0 quantity down to max_num_files_to_group.
```
Comment (wiedld) [inline compactor/src/components/file_classifier/split_based.rs:285]: With the introduction of `RoundInfo::CompactRanges`, we have this method now selectively including/removing files for both  `RoundInfo::CompactRanges` and `RoundInfo::ManySmallFiles`. So now any changes in this method impact both.

It's confusing on why we even have `RoundInfo::CompactRanges` needing to remove additional files, as per it's contract is:
```
CompactRanges are overlapping chains of L0s are less than max_compact_size
```
If we need to also consider total files counts, can we instead as include that logic in the initial compact_ranges decisions?

**Review by wiedld — CHANGES_REQUESTED** [review]
Lots of smarter logic going in, which is great. But we are still blurring the lines -- which makes it harder to reason about. See code.

File: compactor/src/components/round_info_source/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:353]: I think its better together.  The range time checks for L1s that span ranges, and splits them.  So if we split it up, chain_range_times would sometimes return vertical split times.  

I'll rename it, and adjust some comments.  Lets see what you think of that.

File: compactor/src/components/round_info_source/mod.rs (hunk 5 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:398]: no, its not related to the split brain problem.  A single round is basically "_what can we do right now?  ok do that_"  If there's more to do, that's the next round.  

For example, a split always means there will be another round (to compact the output of the split).

File: compactor/src/components/round_info_source/mod.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:312]: > After the last PR, the split_times generated in this methods are hints based upon the density distribution. 

not quite.  The split times returned from `vertical_split_times` are not hints.  They are exactly where splits will occur.  When we need to split but don't have a precise time required the last PR made this function apply hints so the times are more convent to deal with.  But the split times are not hints.  

> We are now have the split_hints here

no, this is not a hint.  This is detecting an L1 that spans compact ranges and forces it to be split exactly where specified, so it doesn't span the ranges.  If the L1 didn't get split, dealing with the two affected regions without breaking invariant would be challenging.  

I'll try to get this function commented better :crossed_fingers:

File: compactor/src/components/split_or_compact/start_level_files_to_split.rs (hunk 6 above)
Comment (Joe-Blount) [reply compactor/src/components/split_or_compact/start_level_files_to_split.rs:363]: This PR adds a use that relies on the order being preserved.  Previously nothing cared about the order.  
To be sure, I just tried adding this to a clean branch and all the tests pass, no changes to insta output.

File: compactor/src/components/file_classifier/split_based.rs (hunk 4 above)
Comment (Joe-Blount) [reply compactor/src/components/file_classifier/split_based.rs:227]: Note that there can be many regions, and the right thing to do within each region will be different.

I've flip-flopped in the last few days on the how this should be handled.  I think what needs to happen is the RoundInfo should always have CompactRegions (maybe just 1), and the CompactReason moves to within each CompactRegion.  But that will be a more disruptive change, and I'm not sure about it yet.  so I'm hoping to let this bit of confusion sit for a bit to see if I still like the idea in  a few days.

File: compactor/src/components/file_classifier/split_based.rs (hunk 8 above)
Comment (Joe-Blount) [reply compactor/src/components/file_classifier/split_based.rs:285]: > So now any changes in this method impact both.

correct.  

the range is a time range known to be less than the max bytes, but a potentially unlimited file count.

This will be cleaner soon.  See above comment about every RoundInfo having range(s) and the reason being per range.  I think that's the right thing to do (and it would make this clearer), but I want to think about it a bit more.

File: compactor/src/components/file_classifier/split_based.rs (hunk 4 above)
Comment (Joe-Blount) [reply compactor/src/components/file_classifier/split_based.rs:227]: > I think what needs to happen is the RoundInfo should always have CompactRegions (maybe just 1)

The more I think about this, the more I like it.  There will be some complications - currently the regions only exit until L0s are gone.  That will need to change.  But will clarify the confusion discussed in this PR and help with some other things.

**Review by wiedld — APPROVED** [review]
